packages/neon/neon/lib/src/pages/login_qrcode.dart

```diff
@@ -0,0 +1,105 @@
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+import 'package:flutter_zxing/flutter_zxing.dart';
+import 'package:neon/src/router.dart';
+import 'package:neon/src/utils/exceptions.dart';
+import 'package:neon/src/utils/login_qrcode.dart';
+import 'package:neon/src/widgets/exception.dart';
+
+class LoginQrcodePage extends StatefulWidget {
+  const LoginQrcodePage({
+    super.key,
+  });
+
+  @override
+  State<LoginQrcodePage> createState() => _LoginQrcodePageState();
+}
+
+class _LoginQrcodePageState extends State<LoginQrcodePage> {
+  String? _lastErrorURL;
+
+  @override
+  Widget build(final BuildContext context) => Scaffold(
+        appBar: AppBar(),
+        body: ReaderWidget(
+          showGallery: false,
+          showToggleCamera: false,
+          showScannerOverlay: false,
+          tryHarder: true,
+          cropPercent: 0,
+          scanDelaySuccess: const Duration(seconds: 3),
+          onScan: (final code) async {
+            String? url;
+            try {
+              url = code.text;
+              if (url == null) {
+                throw InvalidQrcodeException();
+              }
+              final match = parseLoginQrcodeURL(url);
+              if (match == null) {
+                throw InvalidQrcodeException();
+              }
+              await processLoginQrcode(context, match);
+            } catch (e, s) {
+              if (_lastErrorURL != url) {
+                debugPrint(e.toString());
+                debugPrint(s.toString());
+
+                _lastErrorURL = url;
+                NeonException.showSnackbar(context, e);
+              }
+            }
+          },
+        ),
+      );
+}
+
+class LoginQrcodeIntermediatePage extends StatefulWidget {
+  const LoginQrcodeIntermediatePage({
+    required this.serverURL,
+    required this.loginName,
+    required this.password,
+    super.key,
+  });
+
+  final String serverURL;
+  final String loginName;
+  final String password;
+
+  @override
+  State<LoginQrcodeIntermediatePage> createState() => _LoginQrcodeIntermediatePageState();
+}
+
+class _LoginQrcodeIntermediatePageState extends State<LoginQrcodeIntermediatePage> {
+  @override
+  void initState() {
+    super.initState();
+    WidgetsBinding.instance.addPostFrameCallback((final _) {
+      unawaited(
+        processLoginQrcode(
+          context,
+          LoginQrcode(
+            server: widget.serverURL,
+            user: widget.loginName,
+            password: widget.password,
+          ),
+        ),
+      );
+    });
+  }
+
+  @override
+  Widget build(final BuildContext context) => const SizedBox();
+}
```

Comment:
I couldn't find a better way than having this intermediate page, because in the router it's only possible to return a location and not execute any logic. @Leptopoda do you know if there is any better method for doing this?

Reply:
You can add a redirect callback on a per route basis.
This would be the `LoginQrcodeIntermediateRoute`:

```dart
@immutable
class LoginQrcodeIntermediateRoute extends GoRouteData {
  const LoginQrcodeIntermediateRoute({
    required this.serverURL,
    required this.loginName,
    required this.password,
  });

  final String serverURL;
  final String loginName;
  final String password;

  @override
  Widget build(final BuildContext context, final GoRouterState state) => const SizedBox();

  @override
  FutureOr<String?> redirect(final BuildContext context, final GoRouterState state) async {
    final result = await LoginCheckServerStatusRoute(serverURL: serverURL).push<bool>(context);
    return LoginCheckAccountRoute(
      serverURL: serverURL,
      loginName: loginName,
      password: password,
    ).location;
  }
}
```

Note that the checkstatus page won't work like this and needs something different :)
I just copy pasted it from the widget.

Reply:
I'm sorry. Looks like you already tried this.
But I think all the logic should be in the routing and not this intermediate page.

Maybe the `LoginQrcodeIntermediateRoute` can actually build the `LoginCheckServerStatusPage`. This page could be adapted to also take the `LoginQrcode` directly (separate named cunstructor?) and if the full qrcode is present (rather than only the server) it will redirect differently when clicking next

Reply:
> Maybe the LoginQrcodeIntermediateRoute can actually build the LoginCheckServerStatusPage

I don't know if I like this. I already tried something similar but it doesn't really improve the overall structure or logic.

I was thinking to unify the LoginCheckServerStatusPage and LoginCheckAccountPage and just let it check the relevant info based on the parameters. Then we can just process the whole QR code on one page without having any additional redirects etc. We could still keep them as separate routes or unify them too, not sure if that improves anything. What do you think?

Reply:
I don't have anything against multiple routes for the same page.
This might even be preferable as the default BackButton would redirect to the correct page.

I don't know how much unifying them makes sense.
I mean there is a reason they are separated in our flow diagram (https://github.com/provokateurin/nextcloud-neon/blob/main/docs/login.md)
I'm also not 100% conviced by the current approach of pushing the LoginCheckServerStatusRoute as that isn't the declarative approach go_router enables us to use (It's generally quite easy to forget a).

Maybe something like this makes sense?
```dart
@TypedGoRoute<LoginRoute>(
  path: '/login',
  name: 'login',
  routes: [
    TypedGoRoute<LoginQrcodeRoute>(
      path: 'qrcode',
      name: 'loginQrcode',
      routes: [
        TypedGoRoute<LoginQrcodeCheckServerStatusRoute>( // LoginCheckServerStatusPage
          path: 'check/server',
        ),
      ],
    ),
    TypedGoRoute<LoginFlowRoute>(
      path: 'flow/:serverURL',
      name: 'loginFlow',
      routes: [
        TypedGoRoute<LoginCheckServerStatusRoute>( // LoginCheckServerStatusPage
          path: 'check/server',
        ),
      ],
    ),
    TypedGoRoute<LoginCheckAccountRoute>(
      path: 'check/account',
      name: 'checkAccount',
    ),
  ],
)
```